pkg/controllers/v1alpha1/thinruntime/thinruntime_controller.go

```diff
@@ -129,7 +129,7 @@ func (r *ThinRuntimeReconciler) Reconcile(context context.Context, req ctrl.Requ
 // SetupWithManager sets up the controller with the Manager.
 func (r *ThinRuntimeReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options, eventDriven bool) error {
 	if eventDriven {
-		return watch.SetupWatcherWithReconciler(mgr, options, r)
+		return watch.SetupWatcherForReconcilerWithDataset(mgr, options, r, common.ThinRuntime)
```

Comment:
How about using the following ways?

```golang

func SetupWatcherWithReconciler(mgr ctrl.Manager, options controller.Options, r Controller, addons map[string]func(Controller) error) error {
	if addons != nil {
		for name, addonFunc := range addons {
			err := addonFunc(r)
			if err != nil {
				return fmt.Errorf("Error executing function %s: %v", name, err)
			}
		}
	}
	// your other code...
}
```

Reply:
I found that `Controller` should also be refactored, it has comments like:
```
// This controller will be moved to RuntimeReconcilerInterface later
type Controller interface {
	// Reconciler performs a full reconciliation for the object referred to by the Request.
	// The Controller will requeue the Request to be processed again if an error is non-nil or
	// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
	Reconcile(context context.Context, req ctrl.Request) (ctrl.Result, error)

	// ControllerName gets the name of controller
	ControllerName() string

	// ManagedResource which is handled by controller
	ManagedResource() client.Object
}
```
There are two concepts being mix-used in Fluid's code: Reconciler, Controller. I would like to refactor this in next PR.

Reply:
Ok